src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -84,112 +84,43 @@ private static unsafe int CompareOrdinalHelper(string strA, string strB)
                 "For performance reasons, callers of this method should " +
                 "check/short-circuit beforehand if the first char is the same.");
 
-            int length = Math.Min(strA.Length, strB.Length);
+            // Check if the second chars are different here
+            // The reason we check if _firstChar is different is because
+            // it's the most common case and allows us to avoid a method call
+            // to here.
+            // The reason we check if the second char is different is because
+            // if the first two chars the same we can increment by 4 bytes,
+            // leaving us word-aligned on both 32-bit (12 bytes into the string)
+            // and 64-bit (16 bytes) platforms.
+
+            // For empty strings, the second char will be null due to padding.
+            // The start of the string is the type pointer + string length, which
+            // takes up 8 bytes on 32-bit, 12 on x64. For empty strings the null
+            // terminator immediately follows, leaving us with an object
+            // 10/14 bytes in size. Since everything needs to be a multiple
+            // of 4/8, this will get padded and zeroed out.
```

Comment:
It is interesting that we are not doing any tricky casing like this for string Equals that is presumably much more common operation than string Compare.

Reply:
filed https://github.com/dotnet/runtime/pull/111806 to experiment with microbenchmarks 🙂

Reply:
BTW: is it actually legal to rely on empty padding after \0 in empty strings? (the comment mentions it)

Reply:
It is borderline legal in CoreLib only.

Tricks like that are likely going to break if we ever need to adopt hardware buffer overrun enforcements like Arm MTE.

Reply:
What's the conclusion on this?

I'm trying to keep the invocation structure and trailing byte trick as-is to maintain performance for specially handled cases.